snuba/web/rpc/common.py

```diff
@@ -130,6 +141,83 @@ def attribute_key_to_expression(attr_key: AttributeKey) -> Expression:
     )
 
 
+def apply_virtual_columns(
+    query: Query, virtual_column_contexts: list[VirtualColumnContext]
+) -> None:
+    """Injects virtual column mappings into the clickhouse query. Works with NORMALIZED_COLUMNS on the table or
+    dynamic columns in attr_str
+
+    attr_num not supported because mapping on floats is a bad idea
+
+    Example:
+
+        SELECT
+          project_name AS `project_name`,
+          attr_str['release'] AS `release`,
+          attr_str['sentry.sdk.name'] AS `sentry.sdk.name`,
+        ... rest of query
+
+        contexts:
+            [   {from_column_name: project_id, to_column_name: project_name, value_map: {1: "sentry", 2: "snuba"}} ]
+
+
+        Query will be transformed into:
+
+        SELECT
+        -- see the project name column transformed and the value mapping injected
+          transform( CAST( project_id, 'String'), array( '1', '2'), array( 'sentry', 'snuba'), 'unknown') AS `project_name`,
```

Comment:
also I think cast should go in other way, if you request a numeric column it'll be better performance if we cast '1' instead of project_id

Reply:
I agree with the performance being better if we cast `'1'` rather than `project_id`. I was doing this to keep the logic and interface simpler